README.md
# peripheral-manager-bluetooth


[![Maintenance](https://img.shields.io/badge/Maintained%3F-yes-green.svg?style=for-the-badge)](https://github.com/nuvlaedge/peripheral-manager-bluetooth/graphs/commit-activity)
[![GitHub issues](https://img.shields.io/github/issues/nuvlaedge/peripheral-manager-bluetooth?style=for-the-badge&logo=github&logoColor=white)](https://GitHub.com/nuvlaedge/peripheral-manager-bluetooth/issues/)
[![Docker pulls](https://img.shields.io/docker/pulls/nuvlaedge/peripheral-manager-bluetooth?style=for-the-badge&logo=Docker&logoColor=white)](https://cloud.docker.com/u/nuvlaedge/repository/docker/nuvlaedge/peripheral-manager-bluetooth)
[![Docker image size](https://img.shields.io/docker/image-size/nuvladev/peripheral-manager-bluetooth/master?style=for-the-badge&logo=docker&logoColor=white)](https://cloud.docker.com/u/nuvlaedge/repository/docker/nuvlaedge/peripheral-manager-bluetooth)


![CI Build](https://github.com/nuvlaedge/peripheral-manager-bluetooth/actions/workflows/main.yml/badge.svg)
![CI Release](https://github.com/nuvlaedge/peripheral-manager-bluetooth/actions/workflows/release.yml/badge.svg)


**This repository contains the source code for the NuvlaEdge Peripheral Manager for Bluetooth devices - this microservice is responsible for the discovery, categorization and management of all [NuvlaEdge](https://sixsq.com/nuvlaedge) Bluetooth peripherals.**

This microservice is an integral component of the NuvlaEdge Engine.


---

**NOTE:** this microservice is part of a loosely coupled architecture, thus when deployed by itself, it might not provide all of its functionalities. Please refer to https://github.com/nuvlaedge/deployment for a fully functional deployment

---

## Build the NuvlaEdge Peripheral Manager for Bluetooth devices

This repository is already linked with Travis CI, so with every commit, a new Docker image is released.

There is a [POM file](pom.xml) which is responsible for handling the multi-architecture and stage-specific builds.

**If you're developing and testing locally in your own machine**, simply run `docker build .` or even deploy the microservice via the local [compose files](docker-compose.yml) to have your changes built into a new Docker image, and saved into your local filesystem.

**If you're developing in a non-master branch**, please push your changes to the respective branch, and wait for Travis CI to finish the automated build. You'll find your Docker image in the [nuvladev](https://hub.docker.com/u/nuvladev) organization in Docker hub, names as _nuvladev/peripheral-manager-bluetooth:\<branch\>_.

## Deploy the NuvlaEdge Peripheral Manager for Bluetooth devices

The NuvlaEdge Peripheral Manager for Bluetooth will only work if a [Nuvla](https://github.com/nuvla/deployment) endpoint is provided and a NuvlaEdge has been added in Nuvla.

Why? Because this microservice has been built to report directly to Nuvla. Every Bluetooth device will be registered in Nuvla and associated with **an existing** NuvlaEdge.

### Prerequisites

 - *Docker (version 18 or higher)*
 - *Docker Compose (version 1.23.2 or higher)*
 - *Linux*

### Environment variables

|                          	|                                                                                                                                                       	|
|-------------------------	|------------------------------------------------------------------------------------------------------------------------------------------------------	|
|           NUVLAEDGE_UUID 	| (**required**) before starting the microservice, make sure you export the ID of the NuvlaEdge you've created through Nuvla: `export NUVLAEDGE_UUID=<nuvlaedge id from nuvla>` 	|
| NUVLA_ENDPOINT_INSECURE 	| if you're using an insecure Nuvla endpoint, set this to `True`: `export NUVLA_ENDPOINT_INSECURE=True`                                                	|
|          NUVLA_ENDPOINT 	| if you're not using [nuvla.io](https://nuvla.io) then set this to your Nuvla endpoint: `export NUVLA_ENDPOINT=<your endpoint>`                                      	|
| | |

### Launching the NuvlaEdge Peripheral Manager for Bluetooth devices

Simply run `docker-compose up --build`

### If Nuvla is running on `localhost`

Simply run `docker-compose -f docker-compose.localhost.yml up --build`

## Testing

This microservice is completely automated, meaning that as long as all the proper environment variables have been correctly set and the right dependencies have been met, the respective Docker container will start by itself and automatically start registering peripherals into Nuvla, in real-time.

## Contributing

This is an open-source project, so all community contributions are more than welcome. Please read [CONTRIBUTING.md](CONTRIBUTING.md)

## Copyright

Copyright &copy; 2021, SixSq SA


CHANGELOG.md
# Changelog
## [1.4.0] - 2023-04-27
### Added
 - Nuvlaedge-common library integration
### Changed
## [1.3.0] - 2022-09-01
### Added
### Changed
 - Migration to NuvlaEdge
## [1.2.0] - 2022-07-18
### Added
### Changed
 - Use common base image for all NE components
 - Downgraded from python 3.9 to 3.8 to synchronize all components
## [1.1.0] - 2021-07-26
### Added 
 - add support for execution in Kubernetes
### Changed
 - use alpine 3.12
## [1.0.0] - 2021-01-07
### Added 
 - peripheral persistency over restarts
### Changed
 - removed nuvla-api
 - fixed deletion problem on last BT device standing
## [0.1.0] - 2020-12-04
        ### Added 
                  - persist required environment configuration
        ### Changed
                  - minor bug fix
## [0.0.3] - 2020-11-19
### Added 
- capitalize peripheral interface
### Changed
## [0.0.2] - 2020-11-12
### Added 
- add metadata to Docker image 
### Changed
- fix bug on peripheral deletion
## [0.0.1] - 2020-10-22
### Added 
- first BT discovery release


code/tests/test_bluetooth.py
import mock
import unittest

from bluetooth_manager.bluetooth_manager import main


class TestBluetooth(unittest.TestCase):

    @unittest.skip("dummy test")
    def test_main(self):
        """Dummy test."""
        main = mock.Mock()
        main.return_value = None
        assert None is main()


code/nuvla.py
import requests
import os
import sys
import json
import http.cookiejar as cookiejar

def login_push(session_template, url, cookie_file):
    return requests.post(url, headers={'content-type':'application/json', 'accept':'application/json'}, cookies=cookie_file, data=session_template)

def login(activated_path, nuvla_endpoint, cookies_file, extra_flags):

    activated = json.load(open(activated_path))
    api_key = activated['api-key']
    secret_key = activated['secret-key']
    if cookies_file:
        cookie = cookiejar.LWPCookieJar(filename=cookies_file)
        print(cookie)
    else:
        cookie = ''
    # Then we have to login first
    session_template = {
        "template": {
            "href": "session-template/api-key",
            "key": api_key,
            "secret": secret_key
        }
    }

    url = '{}/api/session'.format(nuvla_endpoint)

    if cookies_file:
        session_alive = requests.get(url, headers={'content-type':'application/json'}, cookies=cookie)
        print(session_alive)
        if session_alive:
            print(login_push(session_template, url, cookie))
    else:
        print(login_push(session_template, url, cookie))

def add_bluetooth_peripheral(put_json, nuvla_endpoint, nuvla_id, activated_path, cookies_file, extra_flags):
   
    login(activated_path, nuvla_endpoint, cookies_file, extra_flags)

    url = '{}/api/{}'.format(nuvla_endpoint, nuvla_id) 
    r = requests.put(url, cookies=cookies_file, headers={'content-type':'application/json', 'accept': 'application/json'}, data=put_json)

def remove_bluetooth_peripheral(resource_id, nuvla_endpoint, data, activated_path, cookies_file, extra_flags):
    
    login(activated_path, nuvla_endpoint, cookies_file, extra_flags)

    url = '{}/api/{}'.format(nuvla_endpoint, resource_id) 
    r = requests.delete(url, headers={'content-type':'application/json'}, cookies=cookies_file)




if __name__ == "__main__":
    activated_path = '/home/quietswami/nuvlabox/shared/.activated'
    cookies_file = '/home/quietswami/nuvlabox/shared/cookies'

    login(activated_path, 'http://nuvla.io/', cookies_file, {})

code/manager2.py
#!/usr/bin/env python3

# -*- coding: utf-8 -*-

"""NuvlaBox Peripheral Manager Bluetooth

This service provides bluetooth device discovery.

"""

from bluepy.btle import Scanner, DefaultDelegate
import logging
import requests
import sys
import time
from threading import Event
import json
import converter


def init_logger():
    """ Initializes logging """

    root = logging.getLogger()
    root.setLevel(logging.DEBUG)

    handler = logging.StreamHandler(sys.stdout)
    handler.setLevel(logging.DEBUG)
    formatter = logging.Formatter('%(levelname)s - %(funcName)s - %(message)s')
    handler.setFormatter(formatter)
    root.addHandler(handler)


def wait_bootstrap(healthcheck_endpoint="http://agent/api/healthcheck"):
    """ Simply waits for the NuvlaBox to finish bootstrapping, by pinging the Agent API
    :returns
    """

    logging.info("Checking if NuvlaBox has been initialized...")

    r = requests.get(healthcheck_endpoint)
    
    while not r.ok:
        time.sleep(5)
        r = requests.get(healthcheck_endpoint)

    logging.info('NuvlaBox has been initialized.')
    return

def publish(url, assets):
    """
    API publishing function.
    """

    x = requests.post(url, json=assets)
    return x.json()


def send(url, assets):
    """ Sends POST request for registering new peripheral """

    logging.info("Sending Bluetooth Device information to Nuvla")
    return publish(url, assets)


def remove(url, assets):
    logging.info("Removing Bluetooth Device from Nuvla")
    x = requests.delete(url, json=assets)
    return x.json()

class ScanDelegate(DefaultDelegate):
    def __init__(self):
        DefaultDelegate.__init__(self)

    def handleDiscovery(self, dev, isNewDev, isNewData):
        d = {
            "available": True,
            "name": "",
            "classes": [],
            "identifier": "bluetooth-le",
            "interface": ""
        }

        if isNewDev:
            d['interface'] = dev.addr
            d['available'] = dev.connectable
            data = dev.getScanData()
            for i in data:
                if i[0] == 9:
                    d['name'] = i[-1]
                elif i[0] == 13:
                    d['classes'] = converter.convert(i[-1])
                
            # send(API_URL, d)
            # print(d)
        else:
            d['interface'] = dev.addr
            d['available'] = dev.connectable
            data = dev.getScanData()
            for i in data:
                if i[0] == 9:
                    d['name'] = i[-1]
                elif i[0] == 13:
                    d['classes'] = converter.convert(i[-1])
            print(d)    
if __name__ == "__main__":

    init_logger()

    API_BASE_URL = "http://agent/api"

    # wait_bootstrap()

    API_URL = API_BASE_URL + "/peripheral"

    # e = Event()

    scanner = Scanner().withDelegate(ScanDelegate())

    scanner.start()

    while True:
        scanner.process()





code/converter.py
import re


def convert(number):
    m = re.match('(0x)?([0-9A-Fa-f]{6})', number)
    if m is None:
        print("Invalid class, skipping (%s)" % number)

    hex_string = m.group(2)

    CoD = int(hex_string, 16)

    # Major Device Classes
    classes = ['Miscellaneous', 'Computer', 'Phone',
               'LAN/Network Access Point', 'Audio/Video',
               'Peripheral', 'Imaging', 'Wearable', 'Toy',
               'Health']

    major_number = (CoD >> 8) & 0x1f
    if major_number < len(classes):
        major = classes[major_number]
    elif major_number == 31:
        major = 'Uncategorized'
    else:
        major = 'Reserved'

    # Minor - varies depending on major
    minor_number = (CoD >> 2) & 0x3f
    minor = None

    # computer
    if major_number == 1:
        classes = [
            'Uncategorized', 'Desktop workstation', 'Server-class computer',
            'Laptop', 'Handheld PC/PDA (clamshell)', 'Palm-size PC/PDA',
            'Wearable computer (watch size)', 'Tablet']
        if minor_number < len(classes):
            minor = classes[minor_number]
        else:
            minor = 'reserved'

    # phone
    elif major_number == 2:
        classes = [
            'Uncategorized', 'Cellular', 'Cordless', 'Smartphone',
            'Wired modem or voice gateway', 'Common ISDN access']
        if minor_number < len(classes):
            minor = classes[minor_number]
        else:
            minor = 'reserved'

    # network access point
    elif major_number == 3:
        minor_number >> 3
        classes = [
            'Fully available', '1% to 17% utilized', '17% to 33% utilized',
            '33% to 50% utilized', '50% to 67% utilized',
            '67% to 83% utilized', '83% to 99% utilized',
            'No service available']
        if minor_number < len(classes):
            minor = classes[minor_number]
        else:
            minor = 'reserved'

    # audio/video
    elif major_number == 4:
        classes = [
            'Uncategorized', 'Wearable Headset Device', 'Hands-free Device',
            '(Reserved)', 'Microphone', 'Loudspeaker', 'Headphones',
            'Portable Audio', 'Car audio', 'Set-top box', 'HiFi Audio Device',
            'VCR', 'Video Camera', 'Camcorder', 'Video Monitor',
            'Video Display and Loudspeaker', 'Video Conferencing',
            '(Reserved)', 'Gaming/Toy']
        if minor_number < len(classes):
            minor = classes[minor_number]
        else:
            minor = 'reserved'

    # peripheral
    elif major_number == 5:
        feel_number = minor_number >> 4
        classes = [
            'Not Keyboard / Not Pointing Device', 'Keyboard',
            'Pointing device', 'Combo keyboard/pointing device']
        feel = classes[feel_number]

        classes = [
            'Uncategorized', 'Joystick', 'Gamepad', 'Remote control',
            'Sensing device', 'Digitizer tablet', 'Card Reader', 'Digital Pen',
            'Handheld scanner for bar-codes, RFID, etc.',
            'Handheld gestural input device']
        if minor_number < len(classes):
            minor_low = classes[minor_number]
        else:
            minor_low = 'reserved'

        minor = '%s, %s' % (feel, minor_low)

    # imaging
    elif major_number == 6:
        minors = []
        if minor_number & (1 << 2):
            minors.append('Display')
        if minor_number & (1 << 3):
            minors.append('Camera')
        if minor_number & (1 << 4):
            minors.append('Scanner')
        if minor_number & (1 << 5):
            minors.append('Printer')
        if len(minors > 0):
            minors = ', '.join(minors)

    # wearable
    elif major_number == 7:
        classes = ['Wristwatch', 'Pager', 'Jacket', 'Helmet', 'Glasses']
        if minor_number < len(classes):
            minor = classes[minor_number]
        else:
            minor = 'reserved'

    # toy
    elif major_number == 8:
        classes = ['Robot', 'Vehicle', 'Doll / Action figure', 'Controller',
                   'Game']
        if minor_number < len(classes):
            minor = classes[minor_number]
        else:
            minor = 'reserved'

    # health
    elif major_number == 9:
        classes = [
            'Undefined', 'Blood Pressure Monitor', 'Thermometer',
            'Weighing Scale', 'Glucose Meter', 'Pulse Oximeter',
            'Heart/Pulse Rate Monitor', 'Health Data Display', 'Step Counter',
            'Body Composition Analyzer', 'Peak Flow Monitor',
            'Medication Monitor', 'Knee Prosthesis', 'Ankle Prosthesis',
            'Generic Health Manager', 'Personal Mobility Device']
        if minor_number < len(classes):
            minor = classes[minor_number]
        else:
            minor = 'reserved'

    services = []
    if CoD & (1 << 23):
        services.append('Information')
    if CoD & (1 << 22):
        services.append('Telephony')
    if CoD & (1 << 21):
        services.append('Audio')
    if CoD & (1 << 20):
        services.append('Object Transfer')
    if CoD & (1 << 19):
        services.append('Capturing')
    if CoD & (1 << 18):
        services.append('Rendering')
    if CoD & (1 << 17):
        services.append('Networking')
    if CoD & (1 << 16):
        services.append('Positioning')
    if CoD & (1 << 15):
        services.append('(reserved)')
    if CoD & (1 << 14):
        services.append('(reserved)')
    if CoD & (1 << 13):
        services.append('Limited Discoverable Mode')

    output = [major]
    if minor is not None:
        output.append(' (%s)' % minor)

    return output


if __name__ == "__main__":
    print(convert("0c012a"))


code/manager.py
#!/usr/bin/env python3

# -*- coding: utf-8 -*-

"""NuvlaBox Peripheral Manager Bluetooth
This service provides bluetooth device discovery.
"""

import bluetooth as bt
import logging
import sys
import time
import os
import json
import requests
#from bluetooth.ble import DiscoveryService
from threading import Event


scanning_interval = 30
KUBERNETES_SERVICE_HOST = os.getenv('KUBERNETES_SERVICE_HOST')
namespace = os.getenv('MY_NAMESPACE', 'nuvlabox')


def init_logger():
    """ Initializes logging """

    root = logging.getLogger()
    root.setLevel(logging.DEBUG)

    handler = logging.StreamHandler(sys.stdout)
    handler.setLevel(logging.DEBUG)
    formatter = logging.Formatter('%(levelname)s - %(funcName)s - %(message)s')
    handler.setFormatter(formatter)
    root.addHandler(handler)


def wait_bootstrap(api_url):
    """
    Waits for the NuvlaBox to finish bootstrapping, by checking
        the context file.
    :returns
    """
    while True:
        try:
            logging.info(f'Waiting for {api_url}...')
            r = requests.get(api_url + '/healthcheck')
            r.raise_for_status()
            if r.status_code == 200:
                break
        except:
            time.sleep(15)

    logging.info('NuvlaBox has been initialized.')
    return


def bluetoothCheck(api_url, mac_addr):
    """ Checks if peripheral already exists """
    identifier = mac_addr
    try:
        r = requests.get(f'{api_url}/{identifier}')
        if r.status_code == 404:
            return False
        elif r.status_code == 200:
            return True
        else:
            r.raise_for_status()
    except requests.exceptions.InvalidSchema:
        logging.error(f'The Agent API URL {api_url} seems to be malformed. Cannot continue...')
        raise
    except requests.exceptions.ConnectionError as ex:
        logging.error(f'Cannot reach out to Agent API at {api_url}. Can be a transient issue: {str(ex)}')
        raise
    except requests.exceptions.HTTPError as e:
        logging.warning(f'Could not lookup peripheral {identifier}. Assuming it does not exist')
        return False


def createDeviceFile(device_mac_addr, device_file, peripheral_dir):

    file_path = '{}/{}'.format(peripheral_dir, device_mac_addr)

    with open(file_path, 'w') as outfile:
        json.dump(device_file, outfile)


def removeDeviceFile(device_mac_addr, peripheral_dir):
    file_path = '{}/{}'.format(peripheral_dir, device_mac_addr)

    os.unlink(file_path)


def readDeviceFile(device_mac_addr, peripheral_dir):
    file_path = '{}/{}'.format(peripheral_dir, device_mac_addr)

    return json.load(open(file_path))


def deviceDiscovery():
    """
    Return all discoverable bluetooth devices.
    """
    return bt.discover_devices(lookup_names=True, lookup_class=True)


# def bleDeviceDiscovery():
#     service = DiscoveryService("hci0")
#     devices = service.discover(2)
#     return devices


def compareBluetooth(bluetooth, ble):
    output = []

    for device in bluetooth:
        if device[0] not in ble:
            d = {
                "identifier": device[0],
                "class": device[2],
                "interface": "Bluetooth"
            }

            if device[1] != "":
                d["name"] = device[1]

            output.append(d)

    for device_id, device_name in ble.items():
        d = {
            "identifier": device_id,
            "class": "",    # TODO
            "interface": "Bluetooth-LE"
        }

        if device_name != "":
            d["name"] = device_name

        output.append(d)

    return output


def cod_converter(cod_decimal_string):
    """ From a decimal value of CoD, map and retrieve the corresponding major class of a Bluetooth device

    :param cod_decimal_string: numeric string corresponding to the class of device
    :return: list of class(es)
    """

    if not cod_decimal_string or cod_decimal_string == "":
        return []

    cod_decimal_string = int(cod_decimal_string)

    # Major CoDs
    classes = {0: {'major': 'Miscellaneous',
                     'minor': {}},
               1: {
                   'major': 'Computer',
                   'minor': {
                       'bitwise': False,
                       '0': 'Uncategorized',
                       '1': 'Desktop workstation',
                       '2': 'Server-class computer',
                       '3': 'Laptop',
                       '4': 'Handheld PC/PDA (clamshell)',
                       '5': 'Palm-size PC/PDA',
                       '6': 'Wearable computer (watch size)',
                       '7': 'Tablet'}
               },
               2: {
                   'major': 'Phone',
                   'minor': {
                       'bitwise': False,
                       '0': 'Uncategorized',
                       '1': 'Cellular',
                       '2': 'Cordless',
                       '3': 'Smartphone',
                       '4': 'Wired modem or voice gateway',
                       '5': 'Common ISDN access'
                   }
               },
               3: {
                   'major': 'LAN/Network Access Point',
                   'minor': {
                       'bitwise': False,
                       '0': 'Fully available',
                       '1': '1% to 17% utilized',
                       '2': '17% to 33% utilized',
                       '3': '33% to 50% utilized',
                       '4': '50% to 67% utilized',
                       '5': '67% to 83% utilized',
                       '6': '83% to 99% utilized',
                       '7': 'No service available'
                   }
               },
               4: {
                   'major': 'Audio/Video',
                   'minor': {
                       'bitwise': False,
                       '0': 'Uncategorized',
                       '1': 'Wearable Headset Device',
                       '2': 'Hands-free Device',
                       '3': '(Reserved)',
                       '4': 'Microphone',
                       '5': 'Loudspeaker',
                       '6': 'Headphones',
                       '7': 'Portable Audio',
                       '8': 'Car audio',
                       '9': 'Set-top box',
                       '10': 'HiFi Audio Device',
                       '11': 'VCR',
                       '12': 'Video Camera',
                       '13': 'Camcorder',
                       '14': 'Video Monitor',
                       '15': 'Video Display and Loudspeaker',
                       '16': 'Video Conferencing',
                       '17': '(Reserved)',
                       '18': 'Gaming/Toy'
                   }
               },
               5: {
                   'major': 'Peripheral',
                   'minor': {
                       'bitwise': False,
                       'feel': {
                           '0': 'Not Keyboard / Not Pointing Device',
                           '1': 'Keyboard',
                           '2': 'Pointing device',
                           '3': 'Combo keyboard/pointing device'
                       },
                       '0': 'Uncategorized',
                       '1': 'Joystick',
                       '2': 'Gamepad',
                       '3': 'Remote control',
                       '4': 'Sensing device',
                       '5': 'Digitizer tablet',
                       '6': 'Card Reader',
                       '7': 'Digital Pen',
                       '8': 'Handheld scanner for bar-codes, RFID, etc.',
                       '9': 'Handheld gestural input device'
                   }
               },
               6: {
                   'major': 'Imaging',
                   'minor': {
                       'bitwise': True,
                       '4': 'Display',
                       '8': 'Camera',
                       '16': 'Scanner',
                       '32': 'Printer'
                   }
               },
               7: {
                   'major': 'Wearable',
                   'minor': {
                       'bitwise': False,
                       '0': 'Wristwatch',
                       '1': 'Pager',
                       '2': 'Jacket',
                       '3': 'Helmet',
                       '4': 'Glasses'
                   }
               },
               8: {
                   'major': 'Toy',
                   'minor': {
                       'bitwise': False,
                       '0': 'Robot',
                       '1': 'Vehicle',
                       '2': 'Doll / Action figure',
                       '3': 'Controller',
                       '4': 'Game'
                   }
               },
               9: {
                   'major': 'Health',
                   'minor': {
                       'bitwise': False,
                       '0': 'Undefined',
                       '1': 'Blood Pressure Monitor',
                       '2': 'Thermometer',
                       '3': 'Weighing Scale',
                       '4': 'Glucose Meter',
                       '5': 'Pulse Oximeter',
                       '6': 'Heart/Pulse Rate Monitor',
                       '7': 'Health Data Display',
                       '8': 'Step Counter',
                       '9': 'Body Composition Analyzer',
                       '10': 'Peak Flow Monitor',
                       '11': 'Medication Monitor',
                       '12': 'Knee Prosthesis',
                       '13': 'Ankle Prosthesis',
                       '14': 'Generic Health Manager',
                       '15': 'Personal Mobility Device'
                   }
               }}

    major_number = (cod_decimal_string >> 8) & 0x1f
    minor_number = (cod_decimal_string >> 2) & 0x3f

    minor_class_name = None
    minor = {'minor': {}}
    if major_number == 31:
        major = {'major': 'Uncategorized'}
    else:
        major = classes.get(major_number, {'major': 'Reserved'})
        minor = classes.get(major_number, minor)

    minor_class = minor.get('minor', {})
    if minor_class.get('bitwise', False):
        # i.e. imaging
        for key, value in minor_class.items():
            try:
                # if key is an integer, it is good to be evaluated
                minor_key = int(key)
            except ValueError:
                continue
            except:
                logging.exception("Failed to evaluate minor device class with key %s" % key)
                continue

            if minor_number & minor_key:
                minor_class_name = value
                break
    else:
         minor_class_name = minor_class.get(str(minor_number), 'reserved')

    major_class_name = major.get('major')

    peripheral_classes = [major_class_name, minor_class_name]

    if 'feel' in minor_class:
        feel_number = minor_number >> 4
        feel_class_name = minor_class['feel'].get(str(feel_number))
        if feel_class_name:
            peripheral_classes.append(feel_class_name)

    return peripheral_classes


def bluetoothManager():

    output = {}

    try:
        # list
        bluetoothDevices = deviceDiscovery()
        logging.info(bluetoothDevices)
    except:
        bluetoothDevices = []
        logging.exception("Failed to discover BT devices")

    bleDevices = {}
    # TODO: implement reliable BLE discovery that works for RPi
    # try:
    #     # dict
    #     bleDevices = bleDeviceDiscovery()
    #     logging.info(bleDevices)
    # except:
    #     bleDevices = {}
    #     logging.exception("Failed to discover BLE devices")

    # get formatted list of bt devices [{},...]
    bluetooth = compareBluetooth(bluetoothDevices, bleDevices)
    if len(bluetooth) > 0:
        for device in bluetooth:
            name = device.get("name", "unknown")
            output[device['identifier']] = {
                    "available": True,
                    "name": name,
                    "classes": cod_converter(device.get("class", "")),
                    "identifier": device.get("identifier"),
                    "interface": device.get("interface", "Bluetooth"),
                }

    return output


def diff(before, after):
    enter = []
    leaving = []

    for key in before.keys():
        if key not in after.keys():
            leaving.append(key)
    
    for key in after.keys():
        if key not in before.keys():
            enter.append(key) 

    return enter, leaving


def post_peripheral(api_url: str, body: dict) -> dict:
    """ Posts a new peripheral into Nuvla, via the Agent API

    :param body: content of the peripheral
    :param api_url: URL of the Agent API for peripherals
    :return: Nuvla resource
    """

    try:
        r = requests.post(api_url, json=body)
        r.raise_for_status()
        return r.json()
    except:
        logging.error(f'Cannot create new peripheral in Nuvla. See agent logs for more details on the problem')
        # this will be caught by the calling block
        raise


def delete_peripheral(api_url: str, identifier: str, resource_id=None) -> dict:
    """ Deletes an existing peripheral from Nuvla, via the Agent API

    :param identifier: peripheral identifier (same as local filename)
    :param api_url: URL of the Agent API for peripherals
    :param resource_id: peripheral resource ID in Nuvla
    :return: Nuvla resource
    """

    if resource_id:
        url = f'{api_url}/{identifier}?id={resource_id}'
    else:
        url = f'{api_url}/{identifier}'

    try:
        r = requests.delete(url)
        r.raise_for_status()
        return r.json()
    except:
        logging.error(f'Cannot delete peripheral {identifier} from Nuvla. See agent logs for more info about the issue')
        # this will be caught by the calling block
        raise


def remove_legacy_peripherals(api_url: str, peripherals_dir: str, protocols: list):
    """ In previous versions of this component, the peripherals were stored in an incompatible manner.
    To avoid duplicates, before starting this component, we make sure all legacy peripherals are deleted

    :param api_url: agent api url for peripherals
    :param peripherals_dir: path to peripherals dir
    :param protocols: list of protocols to look for
    :return:
    """

    for proto in protocols:
        if not proto:
            # just to be sure we don't delete the top directory
            continue

        path = f'{peripherals_dir}{proto}'
        if os.path.isdir(path):
            for legacy_peripheral in os.listdir(path):
                with open(f'{path}/{legacy_peripheral}') as lp:
                    nuvla_id = json.load(lp).get("resource_id")

                # if it has a nuvla_id, there it must be removed from Nuvla
                if nuvla_id:
                    try:
                        delete_peripheral(api_url, f"{proto}/{legacy_peripheral}", resource_id=nuvla_id)
                        continue
                    except:
                        pass

                logging.info(f'Removed legacy peripheral {proto}/{legacy_peripheral}. If it still exists, it shall be re-created.')
                os.remove(f'{path}/{legacy_peripheral}')

            # by now, dir must be empty, so this shall work
            os.rmdir(path)
            logging.info(f'Removed all legacy peripherals for interface {proto}: {path}')


def get_saved_peripherals(api_url, protocol):
    """
    To be used at bootstrap, to check for existing peripherals, just to make sure we delete old and only insert new
    peripherals, that have been modified during the NuvlaBox shutdown

    :param api_url: url of the agent api for peripherals
    :param protocol: protocol name = interface
    :return: map of device identifiers and content
    """

    query = f'{api_url}?parameter=interface&value={protocol}'
    r = requests.get(query)
    r.raise_for_status()

    return r.json()


if __name__ == "__main__":

    init_logger()
    logging.info('BLUETOOTH MANAGER STARTED')
    e = Event()

    peripheral_path = '/srv/nuvlabox/shared/.peripherals/'
    agent_api_endpoint = 'localhost:5080' if not KUBERNETES_SERVICE_HOST else f'agent.{namespace}'
    base_api_url = f"http://{agent_api_endpoint}/api"
    API_URL = f"{base_api_url}/peripheral"

    wait_bootstrap(base_api_url)

    remove_legacy_peripherals(API_URL, peripheral_path, ["bluetooth"])

    old_devices = get_saved_peripherals(API_URL, 'Bluetooth')

    while True:

        current_devices = bluetoothManager()
        logging.info('CURRENT DEVICES: {}'.format(current_devices))
        
        if current_devices != old_devices:

            publishing, removing = diff(old_devices, current_devices)

            for device in publishing:

                peripheral_already_registered = bluetoothCheck(API_URL, device)

                if not peripheral_already_registered:

                    logging.info('PUBLISHING: {}'.format(current_devices[device]))
                    try:
                        resource = post_peripheral(API_URL, current_devices[device])
                    except Exception as ex:
                        logging.error(f'Unable to publish peripheral {device}: {str(ex)}')
                        continue

                old_devices[device] = current_devices[device]

            for device in removing:
                
                logging.info('REMOVING: {}'.format(old_devices[device]))

                peripheral_already_registered = bluetoothCheck(API_URL, device)

                if peripheral_already_registered:
                    try:
                        resource = delete_peripheral(API_URL, device)
                    except:
                        logging.exception(f'Cannot delete {device} from Nuvla')
                        continue
                else:
                    logging.warning(f'Peripheral {device} seems to have been removed already')

                del old_devices[device]

        e.wait(timeout=scanning_interval)
